lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/IncrementalSummarizer.java

```diff
@@ -0,0 +1,320 @@
+            }
+        }
+        outlierItemsetPaneCount.keySet().removeAll(unSupported);
+        inlierItemsetPaneCount.keySet().removeAll(unSupported);
+        outlierItemsetWindowCount.keySet().removeAll(unSupported);
+        inlierItemsetWindowCount.keySet().removeAll(unSupported);
+    }
+
+    private void trackItemset(Set<Integer> itemset, double count) {
+        trackingMap.put(itemset, inlierPaneCounts.size() - 1);
+        outlierItemsetWindowCount.put(itemset, count);
+        outlierItemsetPaneCount.put(itemset, count);
+    }
+
+    private void addNewFrequent() {
+        double minSupport = Math.ceil(minOutlierSupport * outlierItemsets.size());
+        HashMap<Integer, Double> inlierPaneSingletonCount = new ExactCount().count(inlierItemsets).getCounts();
+        if (useAttributeCombinations) { // Use attribute combinations
+            // Get new frequent itemsets in outliers
+            FPGrowth fpGrowth = new FPGrowth();
+            List<ItemsetWithCount> frequent = fpGrowth.getItemsetsWithSupportCount(outlierItemsets,
+                    minSupport);
+            List<ItemsetWithCount> newFrequent = new ArrayList<>();
+            for (ItemsetWithCount iwc : frequent) {
+                Set<Integer> itemset = iwc.getItems();
+                if (!outlierItemsetWindowCount.containsKey(itemset)) {
+                    trackItemset(itemset, iwc.getCount());
+                    newFrequent.add(iwc);
+                }
+            }
+            // Get support in the inlier transactions
+            List<ItemsetWithCount> inlierCount = fpGrowth.getCounts(
+                    inlierItemsets, inlierPaneSingletonCount, inlierPaneSingletonCount.keySet(), newFrequent);
+            for (ItemsetWithCount iwc : inlierCount) {
+                inlierItemsetWindowCount.put(iwc.getItems(), iwc.getCount());
+                inlierItemsetPaneCount.put(iwc.getItems(), iwc.getCount());
+            }
+        } else { // Only use singleton attribute
+            HashMap<Integer, Double> outlierPaneSingletonCount = new ExactCount().count(outlierItemsets).getCounts();
+            for (Integer item : outlierPaneSingletonCount.keySet()) {
+                Set<Integer> itemset = new HashSet<>(1);
+                itemset.add(item);
+                if (outlierPaneSingletonCount.get(item) >= minSupport && !outlierItemsetWindowCount.containsKey(itemset)) {
+                    trackItemset(itemset, outlierPaneSingletonCount.get(item));
+                    inlierItemsetWindowCount.put(itemset, inlierPaneSingletonCount.get(item));
+                    inlierItemsetPaneCount.put(itemset, inlierPaneSingletonCount.get(item));
+                }
+            }
+        }
+    }
+
+    @Override
+    public void process(DataFrame df) {
+        // 1. Retire old pane counts if necessary
+        if (inlierPaneCounts.size() == numPanes) { removePaneCounts(); }
+        // 2. Add support counts for the new pane
+        encodeAttributes(df);
```

Comment:
this is destructive; are the semantics for `process` always that dfs consumed should not be reused?